File: leetcode_problems/cpp/lc274_h_index.cpp
#include <vector>
#include <iostream>
#include <algorithm>
using std::vector;

class Solution
{
public:
    int run(vector<int> &citations)
    {
        if (citations.size() == 1)
        {
            if (citations[0] == 0)
            {
                return 0;
            }
            else
            {
                return 1;
            }
        }

        std::sort(citations.begin(), citations.end());

        int index = 0;
        int h_index = 0;

        int cur_int = 0;
        int maxlen = citations.size();
        int maxint = citations[maxlen - 1];

        while (cur_int <= maxint)
        {
            int count = maxlen - index;
            if (cur_int <= count)
            {
                h_index = cur_int;
            }
            else
            {
                break;
            }

            while (cur_int == citations[index] && index < maxlen - 1)
            {
                index++;
            }

            cur_int++;
        }

        return h_index;
    }
};

int main()
{
    vector<int> citations = {0, 0};

    Solution sol;
    auto answer = sol.run(citations);

    std::cout << answer << std::endl;
}

File: leetcode_problems/python/lc274_h_index.py
class Solution:
    def hIndex(
        self,
        citations: list[int],
    ) -> int:
        count = [0] * 1001
        for c in citations:
            count[c] += 1

        index = 0
        for i, c in enumerate(count):
            for _ in range(c):
                citations[index] = i
                index += 1
        citations = citations[::-1]

        h_index = 0
        for i, c in enumerate(citations):
            if i < c:
                h_index = i + 1
            else:
                break
        return h_index


if __name__ == "__main__":
    citations = [3, 0, 6, 1, 5]
    print(Solution().hIndex(citations))
